<?xml version="1.0" ?>
<launch>
  <!-- Launch file for the admittance controller -->

  <!-- Main settings -->
  <arg name="prefix" default=""/>
  <arg name="root_name" default="$(arg prefix)_base_link"/>
  <arg name="tip_name" default="$(arg prefix)_wrist_3_link"/>



  <!-- TOPICS FROM THE LOWER-LEVE CONTROLLER -->

  <!-- Communication with the arm, reading the state of the arm and commanding velocities -->
  <arg name="TOPIC_ARM_STATE" default="/$(arg prefix)/ur5_cartesian_velocity_controller/ee_state" />
  <arg name="TOPIC_ARM_COMMAND" default="/$(arg prefix)/ur5_cartesian_velocity_controller/command_cart_vel" />

  <!-- TOPICS AND SETTINGS FOR ADMITTANCE CONTROLLER -->

  <!-- Communication with the Force/Torque sensor, reading the force/torque -->
  <!-- The wrench is specified in robotiq_ft_frame_id -->
  <arg name="TOPIC_EXTERNAL_WRENCH" default="/$(arg prefix)/robotiq_ft_wrench" />
  <arg name="WRENCH_FILTER_FACTOR" default="0.1" />
  <arg name="FORCE_DEAD_ZONE_THRES" default="5.0" />
  <arg name="TORQUE_DEAD_ZONE_THRES" default="0.5" />

  <!-- An input topic that can be used for control (expected to be in base_link)-->
  <arg name="TOPIC_CONTROL_WRENCH" default="/$(arg prefix)/admittance_control/wrench_control" />

  <!-- An input topic that can be used to set the equilibrium point -->
  <arg name="TOPIC_EQUILIBRIUM_DESIRED" default="/$(arg prefix)/admittance_control/equilibrium_desired" />

  <!-- An output topic that can be used to visualize equilibrium point (real equilibrium might differ from the desired one)-->
  <arg name="TOPIC_EQUILIBRIUM_REAL" default="/$(arg prefix)/admittance_control/equilibrium_real" />

  <!-- An input topic to receive desired velocity from DS only the linear motion-->
  <arg name="TOPIC_DS_VELOCITY" default="/$(arg prefix)/ds_desired_velocity" />


  <!-- The received wrench by the FT-sensor and the controller will be published for debugging/visualization -->
  <!-- They will be specified in base_link -->
  <arg name="TOPIC_EXTERNAL_WRENCH_ARM_FRAME" default="/admittance_control/transformed/external_wrench" />
  <arg name="TOPIC_CONTROL_WRENCH_ARM_FRAME" default="/admittance_control/transformed/control_wrench" />

  <!-- admittance ratio is a value between zero and one that modulated the input force as to control the stiffness-->
  <arg name="TOPIC_AMITTANCE_RATIO" default="/admittance_control/admittance_ratio" />


  <!-- The node will publish the position and the velocity of the in the world frame (Not very sure, need to be checked) -->
  <arg name="TOPIC_ARM_POSE_WORLD" default="/$(arg prefix)/admittance_control/ee_pose_world" />
  <arg name="TOPIC_ARM_TWIST_WORLD" default="/$(arg prefix)/admittance_control/ee_twist_world" />


  <!-- Loading the admittance parameters from a yaml file -->
  <arg name="ADMITTANCE_PARAMS_FILE" default="$(find admittance_control)/config/admittance_params_real.yaml" />

  <group ns="admittance_control">
    <!-- Converting args into params -->
    <param name="topic_arm_state" value="$(arg TOPIC_ARM_STATE)" />
    <param name="topic_arm_command" value="$(arg TOPIC_ARM_COMMAND)" />
    <param name="topic_external_wrench" value="$(arg TOPIC_EXTERNAL_WRENCH)" />
    <param name="topic_control_wrench" value="$(arg TOPIC_CONTROL_WRENCH)" />
    <param name="topic_equilibrium_desired" value="$(arg TOPIC_EQUILIBRIUM_DESIRED)" />
    <param name="topic_equilibrium_real" value="$(arg TOPIC_EQUILIBRIUM_REAL)" />
    <param name="topic_ds_velocity" value="$(arg TOPIC_DS_VELOCITY)" />
    <param name="topic_external_wrench_arm_frame" value="$(arg TOPIC_EXTERNAL_WRENCH_ARM_FRAME)"/>
    <param name="topic_control_external_arm_frame" value="$(arg TOPIC_CONTROL_WRENCH_ARM_FRAME)" />
    <param name="topic_admittance_ratio" value="$(arg TOPIC_AMITTANCE_RATIO)" />
    <param name="topic_arm_pose_world" value="$(arg TOPIC_ARM_POSE_WORLD)" />
    <param name="topic_arm_twist_world" value="$(arg TOPIC_ARM_TWIST_WORLD)" />
    <param name="wrench_filter_factor" value="$(arg WRENCH_FILTER_FACTOR)" />
    <param name="force_dead_zone_thres" value="$(arg FORCE_DEAD_ZONE_THRES)" />
    <param name="torque_dead_zone_thres" value="$(arg TORQUE_DEAD_ZONE_THRES)" />
    <param name="prefix" value="$(arg prefix)" />
    <param name="root_name" value="$(arg root_name)" />
    <param name="tip_name" value="$(arg tip_name)" />

    <rosparam file="$(arg ADMITTANCE_PARAMS_FILE)" command="load"/>
    <node name="admittance_controller" pkg="admittance_control" type="admittance_controller_node" output="screen"/>

  </group>


</launch>
